### PR TITLE
Feat autograd engine patch

### PR DESCRIPTION
**Pull request by wyg1997**

### CODE REVIEW

File: oneflow/api/python/autograd/autograd.cpp [hunk 1]
```diff
@@ -18,23 +18,45 @@ limitations under the License.
...
+// If output is a scalar tensor, out_grad will also be a scaler or empty(will be inited to
 // `flow.ones([1])`).
 Maybe<one::TensorTuple> CheckAndInitOutGrads(const one::TensorTuple& outputs,
                                              const one::TensorTuple& out_grads) {
   auto gradients = std::make_shared<one::TensorTuple>(out_grads.size());
-  TODO();  // Wang Yinggang. Checks all out_grads and push default value for empty item
+  for (int i = 0; i < out_grads.size(); ++i) {
+    if (out_grads.at(i).get()) {
+      CHECK_OR_RETURN(*(outputs.at(i)->shape()) == *(out_grads.at(i)->shape()))
+          << "out_grad's shape must be same as output's";
```
Comment (hjchen2) [inline oneflow/api/python/autograd/autograd.cpp:48]: << "The out_grad shape  must be same as output. (" << outputs.at(i)->shape()->ToString() << " vs " << out_grads.at(i)->shape()-> ToString() << ")";

File: oneflow/api/python/autograd/autograd.cpp [hunk 2]
```diff
@@ -54,6 +76,10 @@ Maybe<one::TensorTuple> Grad(const one::TensorTuple& outputs, const one::TensorT
                              bool create_graph) {
   if (create_graph) { retain_graph = true; }
   if (inputs.empty()) { return Backward(outputs, out_grads, retain_graph, create_graph); }
+  CHECK_OR_RETURN(std::all_of(
+      inputs.begin(), inputs.end(),
+      [](const std::shared_ptr<one::Tensor>& tensor) { return tensor->requires_grad(); }))
+      << "All inputs tensor `.requires_grad` will be true";
```
Comment (hjchen2) [inline oneflow/api/python/autograd/autograd.cpp:82]: will -> should

**Review by hjchen2 — APPROVED** [review]

File: oneflow/core/autograd/autograd_engine.cpp [hunk 3]
```diff
@@ -30,21 +35,39 @@ bool IsReadyToRun(const std::vector<std::shared_ptr<TensorArg>>& out_grads) {
...
     }
   }
   return Maybe<void>::Ok();
 }
 
+Maybe<void> CopyOrAccGrad(Tensor* tensor, bool autograd_mode) {
+  autograd::AutoGradMode mode(autograd_mode);
+  const auto& tensor_arg = tensor->now_grad_arg();
+  if (tensor->acc_grad()) {
+    TensorTuple input = {tensor->acc_grad(), tensor_arg->GetAccTensor().GetPtrOrThrow()};
```
[the remark this answers is not in the document]
Comment (wyg1997) [reply oneflow/core/autograd/autograd_engine.cpp:55]: 哦是的，这里不应该中断

File: oneflow/core/framework/op_expr_helper.cpp [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+namespace oneflow {
+
+namespace op_expr_helper {
+
+Maybe<one::UserOpExpr> AddNOp(int32_t n) {
+  return one::OpBuilder("add_n").Input("in", n).Output("out").Build();
+}
+
+Maybe<one::UserOpExpr> AddOp() {
+  return one::OpBuilder("add").Input("in", 2).Output("out").Build();
```
Comment (hjchen2) [inline oneflow/core/framework/op_expr_helper.cpp:43]: 好像没有add这个op，这里指定op type为"add"是不对的。
one::OpBuilder("add_n").Input("in", 2).Output("out").Build();

**Pull request merged**